core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -228,15 +230,25 @@ class ReplicaFetcherThread(name: String,
 
   override def buildFetchRequest(partitionMap: Seq[(TopicPartition, PartitionFetchState)]): FetchRequest = {
     val requestMap = new util.LinkedHashMap[TopicPartition, JFetchRequest.PartitionData]
+    val partitionsWithError = mutable.Set[TopicPartition]()
 
     partitionMap.foreach { case (topicPartition, partitionFetchState) =>
       // We will not include a replica in the fetch request if it should be throttled.
       if (partitionFetchState.isReadyForFetch && !shouldFollowerThrottle(quota, topicPartition)) {
-        val logStartOffset = replicaMgr.getReplicaOrException(topicPartition).logStartOffset
-        requestMap.put(topicPartition, new JFetchRequest.PartitionData(partitionFetchState.fetchOffset, logStartOffset, fetchSize))
+        try {
+          val logStartOffset = replicaMgr.getReplicaOrException(topicPartition).logStartOffset
+          requestMap.put(topicPartition, new JFetchRequest.PartitionData(partitionFetchState.fetchOffset, logStartOffset, fetchSize))
+        } catch {
+          case e: KafkaStorageException =>
+            partitionsWithError += topicPartition
+            partitionStates.moveToEnd(topicPartition)
+        }
       }
     }
 
+    if (partitionsWithError.nonEmpty)
+      handlePartitionsWithErrors(partitionsWithError)
```

Comment:
Organization wise, it seems it will be cleaner if we return a set of error partitions to the caller and let AbstractFetcherThread call handlePartitionsWithErrors(). Otherwise, both AbstractFetcherThread and ReplicaFetcherThread can call handlePartitionsWithErrors() and it may be harder to track things down.

Reply:
It seems that every time ReplicaFetcherThread calls handlePartitionsWithErrors(), the partitions that have error will not be returned to the caller. In other words the partitions will error has been filtered from the value returned to the caller. In most cases this should prevent caller from calling handlePartitionsWithErrors() for a partition more than once.

I agree that we need to return the partition with error to caller to make sure handlePartitionsWithErrors() is invoked for a given partition at most once. In order to do this we need to modify the buildFetchRequest(), maybeTruncate() and buildLeaderEpochRequest() to return the current return value and the partitions with error as 2-tuple. The code will be a bit messy though. Do you think we should do it?

Reply:
Perhaps we could just add a new class that includes the fetch request and the error partitions and use that as the return type?

Reply:
@junrao Thanks for the comment. If we were to return error partitions in the return value of `buildFetchRequest()`, we probably also want to similarly change the return type of `maybeTruncate()` and `buildLeaderEpochRequest()`, which also call handlePartitionsWithError currently. We will need to change interface definition of these three methods in AbstractFetcherThread and all their implementation. If we don't use 2-tuple, we will need to define three case classes.

The change will make the code a big uglier than it is. So I am trying to understand whether it is worthwhile to do it now. I am inclined to not add this complexity for the following two reasons:

Firstly, it seems that there is no realistic benefit of doing handlePartitionsWithError() only in the AbstractFetcherThread. Currently handlePartitionsWithError is idempotent and it is harmless to call it multiple times for the same partition. We can make the change in the future if handlePartitionsWithError() is no longer idempotent.

Secondly, currently we already ensures that AbstractFetcherThread will not call handleParitionsWithError for the same partition again if it has been called in the ReplicaFetcherThread. This is ensured by not returning error partitions in the response. For example, the FetchRequest returned by buildFetchRequest() will not include the error partition, and therefore this partition won't be processed in the following stages (e.g. processPartitionData). It may be simpler to just keep this logic going forward.

What do you think?

Reply:
I agree with Jun that centralising side-effects makes the code easier to reason about. If you have multiple places where you need to return an existing return value and the error partitions, you only need one case class:

```scala
case class ResultWithErrors[R](result: R, partitionsWithErrors: Set[TopicPartition])
```

The second field could also be a map or whatever else is needed.

Reply:
@ijuma Ah I see. This is nice. I will have the patch to do it.